relay-redis/src/real.rs

```diff
@@ -382,3 +396,154 @@ pub struct Stats {
+    /// Represents a connection pool to a single redis server.
+    Single(
+        bb8::Pool<bb8_redis::RedisConnectionManager>,
+        RedisConfigOptions,
+    ),
+}
+
+impl AsyncRedisPool {
+    /// Takes a command and executes it on redis.
+    pub async fn query_async<T: FromRedisValue>(&self, cmd: Cmd) -> Result<T, RedisError> {
+        match self {
+            Self::Cluster(pool, options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                timeout(
+                    Duration::from_secs(options.read_timeout),
+                    cmd.query_async(&mut *conn),
+                )
+                .await
+                .map_err(RedisError::AsyncTimeout)?
+                .map_err(RedisError::Redis)
+            }
+            Self::Single(pool, options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                timeout(
+                    Duration::from_secs(options.read_timeout),
+                    cmd.query_async(&mut *conn),
+                )
+                .await
+                .map_err(RedisError::AsyncTimeout)?
+                .map_err(RedisError::Redis)
+            }
+        }
+    }
+
+    /// Forwards the [`Script`] to the pool and loads it asynchronously.
+    /// Returns the SHA1 hash of it.
+    pub async fn load_async(&self, script: &Script) -> Result<String, RedisError> {
+        match self {
+            Self::Cluster(pool, _options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                script
+                    .prepare_invoke()
+                    .load_async(&mut *conn)
+                    .await
+                    .map_err(RedisError::Redis)
+            }
+            Self::Single(pool, _options) => {
+                let mut conn = pool.get().await.map_err(RedisError::AsyncPool)?;
+                script
+                    .prepare_invoke()
+                    .load_async(&mut *conn)
+                    .await
+                    .map_err(RedisError::Redis)
+            }
+        }
+    }
+
+    /// Creates a new cluster based [`AsyncRedisPool`].
+    pub async fn cluster<'a>(
+        servers: impl IntoIterator<Item = &'a str>,
+        opts: &RedisConfigOptions,
+    ) -> Result<Self, RedisError> {
+        let client = ClusterClient::new(servers).map_err(RedisError::Redis)?;
+        let manager = RedisClusterConnectionManager::new(client);
+        let pool = Self::base_pool_builder(opts)
+            .build(manager)
+            .await
+            .map_err(RedisError::Redis)?;
+        Ok(AsyncRedisPool::Cluster(pool, opts.clone()))
+    }
+
+    /// Creates a new [`AsyncRedisPool`] backed by a single redis server.
+    pub async fn single(server: &str, opts: &RedisConfigOptions) -> Result<Self, RedisError> {
+        let manager = bb8_redis::RedisConnectionManager::new(server).map_err(RedisError::Redis)?;
```

Comment:
If we want a custom connection config, I don't think the manager allows us to use [`get_multiplexed_async_connection_with_config`](https://docs.rs/redis/0.27.5/redis/struct.Client.html#method.get_multiplexed_async_connection_with_config). Might need our own one and can think about contributing that upstream separately.

Reply:
I will look into this, seems like a good idea to make the connection manager set the proper config

Reply:
I added another custom ConnectionManager that allows us to use the connection with custom config. It feels more consistent between cluster and non cluster pools because the timeout is set once on pool creation instead of setting it on the connection directly. Thank you for the idea!